### PR TITLE
SRR precipitation changes

### PR DESCRIPTION
**Pull request by kanderso-nrel**
See https://github.com/NREL/rdtools/issues/126#issuecomment-620962653

I still don't fully understand what the SRR code is doing (and won't until I set aside some time to really read through it), but here's a starting point for changing the interaction between detected cleaning events and precipitation.  Instead of handling precipitation behavior at the end when all the intervals are partitioned and processed, this moves it to the partitioning step.  Possible that I've totally missed the mark. 

To do:
- [x] revisit `precip` parameters in docstring to clarify when they are ignored/used
- [x] add a precipitation threshold
- [x] Enable 4 cases for `clean_criterion`: `precip_and_shift`, `precip_or_shift`, `precip`, `shift`
- [x] Update precipitation event definition to be based on a single day, while keeping the association between a precipitation event and a detected cleaning event defined on a three day window 
- [x] Update tests
- [x] Update change log

### CODE REVIEW

**Comment by kanderso-nrel** [comment]
@mdeceglie @matt14muller does this line up with what you were imagining?

Test failures are because I changed a keyword but didn't update the tests, will do that once it's clear that the approach is acceptable.

**Comment by matt14muller** [comment]
Kevin,  looks good and thanks.  It appears that 'precip_or_shift' should be able to handle the case when  precipitation data is empty but I would want to make sure we test this case.  I assume this should get the same results as previously when users didnt bring in precipitaiton data.  I also would like to add the ability for the user to input the minimum length of a dry period for fitting Theil-Sen fits.  Alternatively it could be left alone and this handled in post processing on the output data.

**Comment by mdeceglie** [comment]
@matt14muller Agreed that we should add that ability to specify a minimum interval length. What do you think is the correct behavior for handling those intervals in the Monte Carlo step? The current behavior in this PR is to handle them as "invalid" intervals the same as we would with missing data or intervals with huge uncertainty in the slope. In the Monte Carlo step, these get filled with a random constant SR between the end of the last valid interval and the start of the next.

**Comment by mdeceglie** [comment]
I think this is close. @kanderso-nrel  would you take another look, I can't request you as a formal reviewer because you opened the PR.

File: rdtools/test/soiling_test.py [hunk 1]
```diff
@@ -92,12 +92,27 @@ def test_soiling_srr(normalized_daily, insolation, times):
...
-    assert 0.948867 == pytest.approx(sr, abs=1e-6),\
-        'Soiling ratio with precip_clean_only=True different from expected'
+                                          random_seed=1977, precip=precip, clean_criterion='precip')
+    assert 0.976196 == pytest.approx(sr, abs=1e-6),\
+        "Soiling ratio with clean_criterion='precip' different from expected"
+
+    sr, sr_ci, soiling_info = soiling_srr(normalized_daily, insolation, reps=10,
+                                          random_seed=1977, precip=precip, clean_criterion='shift')
+    assert 0.963133 == pytest.approx(sr, abs=1e-6),\
+        "Soiling ratio with clean_criterion='shift' different from expected"
```
Comment (kanderso-nrel) [inline rdtools/test/soiling_test.py:115]: Would be nice to see all these shared arguments lumped into a dictionary and splatted into each test.  Makes it easier to see what is changing between each test call.  More a general comment, not a blocker here

File: rdtools/soiling.py [hunk 2]
```diff
@@ -130,33 +140,37 @@ def _calc_daily_df(self, day_scale=14, clean_threshold='infer',
...
-        df['clean_event'] = (df.delta > clean_threshold)
-        df['clean_event'] = df.clean_event | out_start | out_end
-        df['clean_event'] = (df.clean_event) & \
-                            (~df.clean_event.shift(-1).fillna(False))
+        df['clean_event_detected'] = (df.delta > clean_threshold)
+        precip_event = (df['precip'] > precip_threshold)
+
+        if clean_criterion == 'precip_and_shift':
+            # Detect which cleaning events are associated with rain within a 3 day window
+            precip_event = precip_event.rolling(3, center=True, min_periods=1).sum() >= 1
```
[the remark this answers is not in the document]
Comment (mdeceglie) [reply rdtools/soiling.py:148]: I agree this is clunky. I think I will change it to `precip_event.rolling(3, center=True, min_periods=1).apply(any).astype(bool)`

**Comment by mdeceglie** [comment]
Great, thanks for testing it @matt14muller. I will incorporate @kanderso-nrel's suggestions, then this will be gtg.

File: rdtools/soiling.py [hunk 3]
```diff
@@ -62,16 +63,25 @@ def _calc_daily_df(self, day_scale=14, clean_threshold='infer',
...
+                default 'shift'
+            The method of partitioning the dataset into soiling intervals.
+            If 'precip_and_shift', only rolling median shifts must coincide
+            with precipitation to be a valid cleaning event.
+            If 'precip_or_shift', rolling median shifts and precipitation
+            events are each sufficient on their own to be a cleaning event.
+            If 'shift', only rolling median shifts are treated as cleaning events.
+            If 'precip', only precipitation events are treated as cleaning events.
+        precip_threshold : float, default 0.01
+            The daily precipitation threshold for defining precipitation cleaning events.
```
[the remark this answers is not in the document]
Comment (mdeceglie) [reply rdtools/soiling.py:84]: Updated to read:
```
precip : pd.Series, default None
Daily total precipitation. Units ambiguous but should be the same as 
precip_threshold. Note default behavior of precip_threshold. (Ignored
if ``clean_criterion='shift'``.)
```
and 

```
precip_threshold : float, default 0.01
The daily precipitation threshold for defining precipitation cleaning events.
Units must be consistent with precip.
```

Hopefully that clarifies.

File: rdtools/soiling.py [hunk 4]
```diff
@@ -62,16 +63,25 @@ def _calc_daily_df(self, day_scale=14, clean_threshold='infer',
...
             threshold
-
         recenter : bool, default True
             Whether to recenter (renormalize) the daily performance to the
             median of the first year
+        clean_criterion : {'precip_and_shift', 'precip_or_shift', 'precip', 'shift'} \
+                default 'shift'
+            The method of partitioning the dataset into soiling intervals.
+            If 'precip_and_shift', only rolling median shifts must coincide
+            with precipitation to be a valid cleaning event.
```
[the remark this answers is not in the document]
Comment (mdeceglie) [reply rdtools/soiling.py:78]: I see what you're saying on clean_criterion, but I can't come up with something I like better without making the parameter name much longer.

**Comment by kanderso-nrel** [comment]
LGTM :+1:

**Pull request merged**